src/logic/wallets/store/reducer/index.ts

```diff
@@ -0,0 +1,61 @@
+import { Action, handleActions } from 'redux-actions'
+
+import { ChainId } from 'src/config/chain.d'
+import { PROVIDER_ACTIONS } from 'src/logic/wallets/store/actions'
+
+export type ProvidersState = {
+  name: string
+  hardwareWallet: boolean
+  smartContractWallet: boolean
+  network: ChainId
+  account: string
+  available: boolean
+  ensDomain: string
+  loaded: boolean
+}
+
+export type ProviderWalletPayload = Pick<ProvidersState, 'name' | 'hardwareWallet' | 'smartContractWallet'>
+export type ProviderNetworkPayload = ProvidersState['network']
+export type ProviderAccountPayload = ProvidersState['account']
+export type ProviderEnsPayload = ProvidersState['ensDomain']
+
+export type ProviderPayloads =
+  | ProviderWalletPayload
+  | ProviderAccountPayload
+  | ProviderNetworkPayload
+  | ProviderEnsPayload
+
+const initialProviderState: ProvidersState = {
+  name: '',
+  hardwareWallet: false,
+  smartContractWallet: false,
+  account: '',
+  network: '',
+  ensDomain: '',
+  available: false,
+  loaded: false,
+}
+
+const providerFactory = (provider: ProvidersState) => {
+  const { name, hardwareWallet, smartContractWallet, account, network } = provider
+  const hasWallet = !!name || !!hardwareWallet || !!smartContractWallet
+  return { ...provider, loaded: hasWallet && !!account && !!network }
+}
+
+export const PROVIDER_REDUCER_ID = 'providers'
+
+const providerReducer = handleActions<ProvidersState, ProviderPayloads>(
+  {
+    [PROVIDER_ACTIONS.WALLET]: (state: ProvidersState, { payload }: Action<ProviderWalletPayload>) =>
+      providerFactory({ ...state, ...payload }),
+    [PROVIDER_ACTIONS.NETWORK]: (state: ProvidersState, { payload }: Action<ProviderNetworkPayload>) =>
+      providerFactory({ ...state, network: payload }),
+    [PROVIDER_ACTIONS.ACCOUNT]: (state: ProvidersState, { payload }: Action<ProviderAccountPayload>) =>
+      providerFactory({ ...state, account: payload, available: !!payload }),
+    [PROVIDER_ACTIONS.ENS]: (state: ProvidersState, { payload }: Action<ProviderEnsPayload>) =>
+      providerFactory({ ...state, ensDomain: payload }),
```

Comment:
Could these actions be merged into one `wallet/updated`? They all seem to be doing the same thing

Reply:
They each refer to different Onboard subscription events.

Reply:
But all these events trigger the same thing (update a property in an object)? So why would they have to be correlated?

Reply:
Would you suggest having the same payload? Some subscriptions only return [part of the provider state](https://docs.blocknative.com/onboard#subscriptions-optional).

Reply:
If they were all one then it would be kind of difficult to follow the dispatched actions as they all refer to differnt Onboard subscriptions. What do you think?

It's possible to combine all the logic into one factory with some payload adjustments though:

```ts
const providerFactory = (state: ProvidersState, payload: Partial<ProvidersState>): ProvidersState => {
  const newProvider = { ...state, ...payload }

  const { name, hardwareWallet, smartContractWallet, account, network } = newProvider

  const hasWallet = !!name || hardwareWallet || smartContractWallet
  return { ...newProvider, loaded: hasWallet && !!account && !!network }
}
```

Reply:
Sorry, I didn't get it. If you want to follow the dispatched action, how would this factory help?

Reply:
I'd say if you want to follow the actions, then the actions are ok

Reply:
The factory example is if we merged all the separate actions into one. I'll leave it be.

Reply:
Confirmed as resolved by @mikheevm